src/main/java/edu/texas/social_computing/hospitals/RHRP.java

```diff
@@ -0,0 +1,134 @@
+                    .get(residentTable.getResidentRankProgress(currentResident)));
+            residentTable.incrementResidentRankProgress(currentResident);
+
+            tryToMatch(hospital,currentResident,residentTable,freeResidents,m);
+
+            // if the currentResident managed to get assigned then try to assign partner if he has one
+            if (m.hasAssignment(currentResident) && currentResident.hasPartner()) {
+                tryToMatch(hospital,residentTable.getResidentById(currentResident.getPartnerId()),
+                        residentTable,freeResidents,m);
+            }
+
+            // if currentResident has more applying to do, put back in Q
+            if (!m.hasAssignment(currentResident)) {
+                freeResidents.add(currentResident);
+                if (currentResident.hasPartner()) {
+                    freeResidents.add(residentTable.getResidentById(currentResident.getPartnerId()));
+                }
+            }
+        }
+        return m;
+    }
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents) {
+        return run(hospitalTable, residentTable, freeResidents, new Matching());
+    }
+
+    public static void main(String[] args) throws FileNotFoundException {
+        List<Resident> residents = FileImporter.importResidents(args[0]);
+        List<Hospital> hospitals = FileImporter.importHospitals(args[1]);
+        System.out.println("loaded Files");
+        System.out.println("residents: " + Integer.toString(residents.size()));
+        System.out.println("hospitals: " + Integer.toString(hospitals.size()));
+
+        // make lookup tables
+        HospitalTable hospitalTable = HospitalTable.create(hospitals);
+        ResidentTable residentTable = ResidentTable.create(residents);
+        System.out.println("generated tables");
+
+        // call HRP
+        Matching m = RHRP.run(hospitalTable, residentTable, new ArrayDeque<>(residentTable.getAll()));
+        System.out.println(m);
+
+
+        Set<String> unassignedIds = m.getAllUnassigned(residentTable.getAll()).stream()
+                .map(Resident::getId)
+                .collect(ImmutableSet.toImmutableSet());
+        System.out.println("Unassigned: " + unassignedIds);
+    }
+
+    private static List<String> getNonRankedIds(Hospital h, List<Resident> residents) {
+        return residents.stream()
+                .filter(res -> h.rankOf(res) == Integer.MAX_VALUE)
+                .map(Resident::getId)
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    private static void tryToMatch(Hospital h, Resident resident,
+                                   ResidentTable residentTable, Queue<Resident> freeResidents, Matching m) {
+        // Only if the hospital has this resident on its pref list can it accept the resident
```

Comment:
I feel like there is some substantial nugget here that is common between this and regular HRP. Can we extract a method in the HRP loop and call it here, too?

Reply:
I am not sure about that b/c with RHRP if the resident has failed to match then the partner's match has to be undone, which is different than our RRHP.

Reply:
right, but I think we should evaluate that outside of `tryToMatch`. I think that `tryToMatch` should do exactly that, and if it fails to find a match, then we need to take care of that outside of this method.